[LifetimeSafety] Add missing origins stats for lifetime analysis

DEBADRIBASAK opened the pull request:
This PR adds the implementation for printing missing origin stats for lifetime analysis.

**Purpose:**

This capability is added to track the expression types with missing origin. While retrieving the origins from origin manager, some expressions show missing origins. Currently these are created on the fly using getOrCreate function. For analysing the coverage of the check, it will be necessary to see what kind of expressions have a missing origin. It prints the counts in this form: `QualType : count` and `StmtClassName : count`.

**Approach:**

1. The signature of the runLifetimeAnalysis function is changed to return the LifetimeAnalysis object which will be used to get the origin manager which can be used for finding the count of missing origins.
2. The count of missing origins is kept in origin manager while the CFG is visited as part of the analysis.

Example output:

For the file llvm-project/llvm/lib/Demangle/Demangle.cpp:

```
*** LifetimeSafety Missing Origin per QualType: (QualType : count) :

value_type : 1
char * : 3


*** LifetimeSafety Missing Origin per StmtClassName: (StmtClassName : count) :

BinaryOperator : 3
UnaryOperator : 1

Total missing origins: 4
```

DEBADRIBASAK committed:
Adding the lifetime stats collection logic to AnalysisBasedWarnings

DEBADRIBASAK committed:
Correcting the signature of getMissingOrigins

DEBADRIBASAK committed:
Merge branch 'main' into lifetime-stats

usx95 reviewed (changes requested):
Thanks for working on this.

One major comment is that this implementation exposes many `internal::` details of lifetime safety to `AnalysisBasedWarnings.h`.

Layering-wise I propose the following changes:
- Define `struct LifetimeSafetyStats { llvm::StringMap<unsigned> MissingOriginCount; }` in `LifetimeSafety.h`
- Have an instance of `LifetimeSafetyStats` as a field of `AnalysisBasedWarnings` instead of `MissingOriginCount`.
- Change `runLifetimeSafetyAnalysis` to accept this stats instance which it can modify if we have collection turned on:
```cpp
void runLifetimeSafetyAnalysis(AnalysisDeclContext &AC, 
                               LifetimeSafetyReporter *Reporter, LifetimeSafetyStats& Stats, bool CollectStats)
```
- Also add a `void clang::lifetimes::PrintStats(const LifetimeSafetyStats& Stats)` in LifetimeSafety.h/.cpp in `clang::lifetimes` namespace which prints the stats to `lvm::errs()`.
- Call `clang::lifetimes::PrintStats(Stats)` from `AnalysisBasedWarnings::PrintStats()`.

DEBADRIBASAK committed:
Adding a LifetimeSafetyStats struct to keep track of missing origins and change the signature of runLifetimeSafetyAnalysis to not return the analysis object

DEBADRIBASAK committed:
Merge branch 'main' into lifetime-stats

usx95 commented on clang/include/clang/Analysis/Analyses/LifetimeSafety/LifetimeSafety.h:
```diff
@@ -82,6 +89,15 @@ class LifetimeSafetyAnalysis {
+void runLifetimeSafetyAnalysis(AnalysisDeclContext &AC,
+                               LifetimeSafetyReporter *Reporter,
+                               LifetimeSafetyStats &Stats, bool CollectStats);
+
```
please keep the public/non-internal functions to the top and before the internal namespace.

usx95 commented on clang/include/clang/Analysis/Analyses/LifetimeSafety/LifetimeSafety.h:
```diff
@@ -23,7 +23,11 @@
+#include "clang/Analysis/Analyses/LifetimeSafety/Origins.h"
 #include "clang/Analysis/AnalysisDeclContext.h"
+#include "llvm/ADT/StringMap.h"
+#include "llvm/Support/raw_ostream.h"
```
IIUC, this looks like an unused header

usx95 commented on clang/lib/Sema/AnalysisBasedWarnings.cpp:
```diff
@@ -52,7 +54,9 @@
 #include "llvm/ADT/StringRef.h"
+#include "llvm/Support/Casting.h"
 #include "llvm/Support/Debug.h"
+#include "llvm/Support/raw_ostream.h"
```
nit: These two look unrelated to changes in this file (although could be not unused).

usx95 reviewed (changes requested):
Awesome. This looks much better now.

One remaining concern I have is that we might be missing to record some missingOrigins here. Suggested a more black box collection style using RecursiveASTVisitor. Also it is important to not do this when `-print-stats`​ is disabled.

DEBADRIBASAK commented on clang/lib/Analysis/LifetimeSafety/Origins.cpp:
```diff
@@ -39,6 +46,17 @@ OriginID OriginManager::get(const Expr &E) {
+    ExprTypeToMissingOriginCount[ExprStr] = 1;
+  } else {
+    CountIt->second++;
+  }
```
Thank you for the feedback. I will incorporate the changes. But I have one query regarding this statement:

> We do this collection at the end of the analysis

Currently we create the origin if it is not available in the map inside `OriginMgr`. If we perform the missing origin count at the end of the analysis it will never show any missing origin.

DEBADRIBASAK replied:
Modified the implementation. For now I am just adding the hasOrigin implementation to Origins.cpp.

DEBADRIBASAK committed:
Moving the missing origin calculation logic after lifetime analysis using an AST visitor

DEBADRIBASAK committed:
Removing an outdated comment

usx95 commented on clang/lib/Sema/AnalysisBasedWarnings.cpp:
```diff
@@ -3159,8 +3164,9 @@ void clang::sema::AnalysisBasedWarnings::IssueWarnings(
 
 void clang::sema::AnalysisBasedWarnings::PrintStats() const {
-  llvm::errs() << "\n*** Analysis Based Warnings Stats:\n";
 
```
This looks unresolved (added the newline somewhere else).

usx95 commented on clang/include/clang/Analysis/Analyses/LifetimeSafety/LifetimeSafety.h:
```diff
@@ -34,6 +38,12 @@ enum class Confidence : uint8_t {
 };
 
+/// A structure to hold the statistics related to LifetimeAnalysis.
+/// Curently it holds only the missing origin details.
```
typo: Currently

usx95 commented on clang/lib/Analysis/LifetimeSafety/Origins.cpp:
```diff
@@ -88,4 +113,33 @@ OriginID OriginManager::getOrCreate(const ValueDecl &D) {
+bool MissingOriginCollector::VisitExpr(Expr *E) {
+  if (!hasOrigin(E))
+    return true;
+  // Check if we have an origin for this expression
```
End sentence with a full stop.

usx95 commented on clang/lib/Sema/AnalysisBasedWarnings.cpp:
```diff
@@ -53,6 +55,7 @@
 #include "llvm/ADT/SmallVector.h"
 #include "llvm/ADT/StringRef.h"
 #include "llvm/Support/Debug.h"
+#include "llvm/Support/raw_ostream.h"
```
nit: remove. This is unsued (in new code).

usx95 commented on clang/lib/Analysis/LifetimeSafety/Origins.cpp:
```diff
@@ -88,4 +113,33 @@ OriginID OriginManager::getOrCreate(const ValueDecl &D) {
+  const auto &ExprToOriginId = OM.getExprToOriginId();
+  auto It = ExprToOriginId.find(E);
+  if (It == ExprToOriginId.end()) {
+    // No origin found - count this as missing
```
nit: end with full stop

usx95 commented on clang/include/clang/Analysis/Analyses/LifetimeSafety/Origins.h:
```diff
@@ -16,7 +16,16 @@
+
+namespace clang::lifetimes {
+
+struct LifetimeSafetyStats;
```
Ah sorry for missing this in my original suggestion. I think you have added this because it creates a cyclic header includes if we include LifetimeSafety.h here

Maybe have a separate header for Stats "clang/include/clang/Analysis/Analyses/LifetimeSafety/Stats.h" `printStats`​ can live in clang/lib/Analysis/LifetimeSafety/Stats.cpp
`collectLifetimeStats`​ would need to still live in LifetimeSafety.cpp

usx95 commented on clang/lib/Analysis/LifetimeSafety/Origins.cpp:
```diff
@@ -7,9 +7,33 @@
+    return RD->hasAttr<PointerAttr>();
+  }
+  return false;
+}
```
FYI: I created a PR https://github.com/llvm/llvm-project/pull/169620 to move this to LifetimeAnnotations.h/.cpp as this also coincides with one of my PRs.

Will let you know if it gets submitted before then we can remove this from here.

usx95 replied:
This has landed. You can remove this after rebasing.

DEBADRIBASAK committed:
Made MissingOriginCollector internal to Origins.cpp

DEBADRIBASAK committed:
Merge branch 'main' into lifetime-stats

DEBADRIBASAK committed:
Introducing a dedicated file for LifetimeSafetyStats and changing the report format.

instead of using ExprClassName<QualType> it will be shown in the following way:

```
*** LifetimeSafety Missing Origin Stats (expression_type : count) :

UnaryOperator (Type: const value_type) : 1
BinaryOperator (Type: char *) : 3
CXXOperatorCallExpr (Type: const value_type) : 2
CXXMemberCallExpr (Type: basic_string_view<char>) : 1
DeclRefExpr (Type: char *) : 9
CXXMemberCallExpr (Type: const_iterator) : 1
CXXOperatorCallExpr (Type: class std::basic_string<char>) : 1
DeclRefExpr (Type: std::string_view) : 27
CXXOperatorCallExpr (Type: basic_string<char>) : 3
Total missing origins: 48
```

Xazax-hun reviewed (commented):
I wonder if collecting in buckets indexed by `StmtClass<QualType>` is the right thing to do. 

If the type and the ast node kind are expected to be correlated, i.e., we expect to have holes where a combination of the type and the kind is the root cause something is not covered than this is the right way to slice the data.

On the other hand, if we expect these to be more independent, i.e., some types are not covered and some kinds of ast nodes are not covered and there is no correlation between them, I think it might be better to collect across these two independent axises.

So print one statistics that maps from statement class (ids, no need for strings there) to counts and print another one that maps from types (again no need for string maps) to counts.

DEBADRIBASAK commented:
> So print one statistics that maps from statement class (ids, no need for strings there) to counts and print another one that maps from types (again no need for string maps) to counts.

[Xazax-hun](https://github.com/Xazax-hun): We had added string maps because it is easier to interpret the types/class names as strings. Will not changing it to ids make it difficult to interpret the report?

usx95 commented on clang/include/clang/Analysis/Analyses/LifetimeSafety/LifetimeStats.h:
```diff
@@ -0,0 +1,33 @@
+//===- LifetimeStats.h - Miscellaneous statistics related to lifetime safety
```
Suggestion to make the heading smaller:

```
//===- LifetimeStats.h - Lifetime Safety Statistics ------------*- C++ -*-===//
```

DEBADRIBASAK committed:
Separating the StmtClassName and QualType for missing origin counts and putting them in separate maps

DEBADRIBASAK commented:
@Xazax-hun @usx95 I do not think it is feasible to add the `StmtClassName` and `QualType` as keys without converting them to strings.

1. I tried using a DenseMap with `QualType` as the keys, but that is not allowed since QualType does not have tombstone.
2. I was planning to use the `StmtClass` as an index in an integer array of size `Stmt::lastStmtConstant+1`, and use `StmtClassInfo` array of `StmtClassNameTable` objects for getting their names while printing. But `StmtClassInfo` is a static array inside `Stmt.cpp` file and we cannot use it in other files.

I used two string maps for now:

```
*** LifetimeSafety Missing Origin per QualType: (QualType : count) :

basic_string<char> : 3
basic_string_view<char> : 1
class std::basic_string<char> : 1
char * : 12
const_iterator : 1
const value_type : 3
std::string_view : 27


*** LifetimeSafety Missing Origin per StmtClassName: (StmtClassName : count) :

CXXOperatorCallExpr : 6
BinaryOperator : 3
CXXMemberCallExpr : 2
UnaryOperator : 1
DeclRefExpr : 36

Total missing origins: 48
```

DEBADRIBASAK committed:
Removing unused headers from AnalysisBasedWarnings

usx95 commented on clang/include/clang/Analysis/Analyses/LifetimeSafety/LifetimeSafety.h:
```diff
@@ -21,9 +21,13 @@
+#include "clang/Analysis/Analyses/LifetimeSafety/Origins.h"
 #include "clang/Analysis/AnalysisDeclContext.h"
+#include "llvm/Support/raw_ostream.h"
+#include <string>
```
Are these somehow getting auto-added by some tool ?

string and raw_ostream are still unused.

usx95 reviewed (approved):
LGTM. 

This looks fine. Thanks for trying out the suggestions. Since this is only populated when `-print-stats`​ is enabled, I am fine with using strings as keys.

Please update the PR description with the latest output.

DEBADRIBASAK committed:
Accepting LifetimeStats instead of individual maps inside Missingorigincollector

usx95 commented:
You could create QualType from this Type* and then print it `Type* T; ..; QualType(T, 0).getAsString();`

DEBADRIBASAK committed:
Modified the LifetimeStats to use constant pointer to Type returned by  in qualType

DEBADRIBASAK committed:
Fomatting changes in variable names

Xazax-hun reviewed (commented):
Is it possible to add a test case? That would also document how to use this feature.

DEBADRIBASAK committed:
Added a LIT test for missing origins stats and put a nullptr check in the caller of collectOriginStats

Xazax-hun reviewed (commented):
Overall looks good to me but there are some conflicts, please resolve those.

DEBADRIBASAK committed:
Merge branch 'main' into lifetime-stats

DEBADRIBASAK commented:
> Overall looks good to me but there are some conflicts, please resolve those.

Resolved conflicts.

Xazax-hun commented on clang/lib/Analysis/LifetimeSafety/Origins.cpp:
```diff
@@ -157,4 +189,20 @@ const Origin &OriginManager::getOrigin(OriginID ID) const {
   return AllOrigins[ID.Value];
 }
 
+OriginID OriginManager::getOrCreate(const ValueDecl &D) {
```
Is this a remnant for a wrong conflict resolution?

DEBADRIBASAK replied:
Yes. Looks like a recent PR changed most of the logic in this file. Corrected the implementation.

DEBADRIBASAK committed:
Modifying the missing origin calculation logic after multi-origin tracking

DEBADRIBASAK committed:
Removing hasOrigin functions from LifetimeAnnotations. This logic is implemented in hasOrigins function

DEBADRIBASAK committed:
Merge branch 'main' into lifetime-stats

usx95 reviewed (approved):
LGTM. Thanks.

Please update the PR description with the new output

DEBADRIBASAK commented:
> LGTM. Thanks.
> 
> Please update the PR description with the new output

Updated the description.

DEBADRIBASAK committed:
Merge branch 'main' into lifetime-stats

DEBADRIBASAK committed:
Formatting changes
